### ISSUE TITLE
Change existing TensorGraph layers to match Keras2 API

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#1095]
We've had a number of technical discussions about ways we can improve `TensorGraph` integration with mainstream TensorFlow. (https://github.com/deepchem/deepchem/issues/1038, https://github.com/deepchem/deepchem/issues/1076, https://github.com/deepchem/deepchem/issues/988). These proposals inevitably run into some serious technical barriers. The main issue is that `TensorGraph` supports a lot of custom constructs (simple multitasking, submodels, serializiation, RNN support, etc) that we need in order to maintain the https://moleculenet.ai models and other advanced models that DeepChem supports. It looks like it's not going to be feasible to implement these models directly in Keras without significant pain.

However, it's interesting to note that Keras (and specifically Keras 2 https://blog.keras.io/introducing-keras-2.html) is more an API specification than a particular library. `tf.keras` is an implementation of this API. The main `keras` library is also an implementation, and I believe that there are others available. Given that `TensorGraph` is already pretty full-featured, I propose we go ahead and turn it into a Keras2 implementation. Here's what this conversion would entail:

- *Changing `TensorGraph` layers to match Keras2 specification*: We've already previously discussed doing this (https://github.com/deepchem/deepchem/issues/988). In general, this will add lots of value since users can move over Keras2 models with set hyperparameters into DeepChem without trouble. This is currently an issue with Dragonn integration (https://github.com/deepchem/deepchem/issues/1002), where it's tricky to port over the Keras2 Dragonn hyperparameters.
- *Add missing Keras2 layers*: This should be straightforward since we can just re-use `tf.keras` layers and need only add a little boilerplate.
- *Add support for functional API*: The Keras functional API (https://keras.io/getting-started/functional-api-guide/) is the blessed way to build complex Keras models. It would be nice to have a `TensorGraph` implementation. (Something like our `Sequential` class that builds the correct `TensorGraph` under the hood would work).

I think that the implementation complexity for these changes should be manage-able. Combined with the `tf.Estimator` support suggested by @peastman (https://github.com/deepchem/deepchem/issues/1076#issuecomment-366386384), I think this set of API changes should be a solid path forward for the next few months.

I'd love feedback on this proposal. For users, does this seem like it would make DeepChem more useful to your day-to-day efforts? For developers, does the proposed implementation path sound feasible?

**Comment by peastman** [comment]
I'm not convinced this is a good idea.  If we want to use Keras, there's already a full implementation in tf.keras.  Creating another complete implementation of the API would be a huge task, and wouldn't really benefit anyone.  And any faithful implementation of the Keras API would have exactly the same problems we ran into in #1038.  So instead we would end up with a halfhearted imitation of Keras that looked similar but worked in fundamentally different ways, and was incompatible with all other implementations.

I think we're still searching for our long term architecture.  My instinct is that it will be either completely based on Keras (which is a nice API, but has limitations that may rule it out), or else Keras will have no part in it.  For example, if we decide to standardize around Estimator, then we should be basing the API for our layers on what's in tf.layers.  (Estimator, IMO, is a *much* less nice API than Keras, but also is unlikely to present fundamental roadblocks.  Sigh.)

**Comment by rbharath** [comment]
These are good points. We don't want to take on unnecessary development effort given that our development team is much smaller than the development teams for Keras/TensorFlow.

At the same time, I think we should at least implement part 1 (*Changing TensorGraph layers to match Keras2 specification*). We already have problems with duplicate layers (https://github.com/deepchem/deepchem/issues/1092#issue-297972942 notes we have two BatchNorm layers). Making sure our standard layers (2d convolution, batch norm, 1d convolutions, etc.) match the Keras2 API will make it much easier for users to move code between DeepChem and Keras.

Agree that we're still looking for our long term architecture. Honestly, the whole field is looking for its long term architecture. My two cents is that pytorch/tf.eager style combined with Keras-like layers is probably the long term winner, but it's going to take years to settle. For now, comparing Keras and tf.Estimator, I agree that Keras has a much nicer API, and just implementing some type of `TensorGraph.make_estimator` function is probably good enough for our `tf.Estimator` integration needs.

**Comment by peastman** [comment]
Wouldn't it be better to change the TensorGraph layers to match the tf.layers specification?  If that's what we're standardizing on.

**Comment by rbharath** [comment]
I think the Keras2 API spec has a lot broader adoption than `tf.layer`. Matching Keras2 will let us move models back and forth between Keras and DeepChem easily.

**Comment by rbharath** [comment]
I've changed the title of issue to match the discussion. If there are no objections, I think we can open this issue out to contributions.

**Comment by rbharath** [comment]
I'm opening this issue out to contributors. If you'd like to work on this issue, pick a layer that `TensorGraph` implements that's also a standard Keras2 layer. For example, a convolution layer or an LSTM layer or the like. Check against the Keras2 API spec, and modify the TensorGraph layer to accept the same arguments and in the same order. Use grep or a similar tool to verify that all uses of the layer in `examples/` or `contrib/` are modified accordingly.

This should be a good starter PR since it requires minimal, mostly mechanical code changes. Note that each contributed PR should target only one Layer in order to allow multiple people to tackle different parts of the task. If you'd like to contribute, please raise a comment on this thread mentioning the layer you'd like to convert and raise a pull request marked [WIP] tracking your work.

**Comment by peastman** [comment]
This won't always be mechanical, and we need to be very careful to avoid breaking things.  To give a simple example, compare the Keras and TensorGraph Dense layers.  Keras has a bunch of options that TensorGraph doesn't include (all the regularizers, for example), so we'll need to add those.  But TensorGraph also supports the `time_series` option that Keras doesn't, so what do we do with that?  Then they have options that do the same thing but with different names: `biases_initializer` and `weights_initializer` in TensorGraph, vs. `bias_initializer` and `kernel_initializer` in Keras.  But it's not just the names that are different: the TensorGraph versions take initializer functions while the Keras versions take strings for an enumerated list of supported initialization methods.  And then of course they have different defaults.  TensorGraph defaults to `variance_scaling_initializer` for the weights while Keras defaults to `'glorot_uniform'`.  So we're going to change the behavior of all existing models that use this layer!

Then there's the GRU and LSTM layers.  Recurrent layers work differently in Keras than in TensorGraph.  Some of the options like `stateful` and `return_state` simply don't make sense, because that isn't how recurrent networks are implemented in TensorGraph.

I'm still very skeptical about whether it's a good idea to make this change.

**Comment by rbharath** [comment]
These are good points. I'm removing the contribution welcome label for the time being. Let's discuss this at the next steering committee meeting to see if there's a reasonable implementation path.

**Comment by rbharath** [comment]
Closing this issue since we've mostly completed the transition to Keras layers.
